### ISSUE TITLE
More Complete Solution for FAPI Internal/External Link Element

### ISSUE THREAD

**Issue by tadean** [az-digital/az_quickstart#647]
## What is the problem that we want to solve?
Some of our Forms and compound Field Widgets have a need to store links, and often these require storage of both internal and external links. This has changed a great deal from D7 and D8-D9 in that there is now a pretty large difference between an absolute link (a string), and an internal route (usually a reference to a specific route and its contained parameters.)

This makes it complicated to create forms that store both kinds of links.

The [FAPI Url element](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Element%21Url.php/class/Url/9.1.x) does not handle internal paths.
The [FAPI Path element](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21Element%21PathElement.php/class/PathElement/9.1.x) does not handle external paths.

Ideally we would like access to some form element that handles both types.

Core sometimes implements this as an Entity autocomplete field that searches internal routes but also has extra validation that allows it to accept extenal links. (see the Menu add link form).

This need is for **custom forms and form elements**, **it is not applicable** to content types that use the **Field Link** type (which already handles both types).

## Conditions of satisfaction
- [ ] Combination of validation/ajax handlers that allow a form element to handle both internal and external links.

**Comment by joeparsons** [comment]
## Background info / Related issues & PRs

Original PR that added custom Card field plugins:
https://github.com/az-digital/az_quickstart/pull/357

Followup tasks for Card module/fields:
https://github.com/az-digital/az_quickstart/issues/377

Cards Paragraph type (utilizes custom Card field plugins):
https://github.com/az-digital/az_quickstart/issues/289
https://github.com/az-digital/az_quickstart/pull/373

Relative paths in Cards link field issue/PR:
https://github.com/az-digital/az_quickstart/issues/537
https://github.com/az-digital/az_quickstart/pull/636

CardWidget plugin improvements:
https://github.com/az-digital/az_quickstart/pull/634
https://github.com/az-digital/az_quickstart/issues/538
https://github.com/az-digital/az_quickstart/issues/427
https://github.com/az-digital/az_quickstart/issues/390
https://github.com/az-digital/az_quickstart/issues/613

**Comment by michaelhagedon** [comment]
OK, I think I see the problem. We're really stretching Drupal internals on this point.

As far as I can tell, everything in Drupal core that does the fancy internal / external link validation (with entity autocomplete) is just using the `link` module, which provides the [Link field widget](https://git.drupalcode.org/project/drupal/-/blob/9.1.10/core/modules/link/src/Plugin/Field/FieldWidget/LinkWidget.php). Obviously that's what you get with content types, but it's also what the menu system uses: menu links are entities, and they also use the link field type: https://git.drupalcode.org/project/drupal/-/blob/9.1.10/core/modules/menu_link_content/src/Entity/MenuLinkContent.php#L326

It sounds to me like ideally what we want is maybe a version of the FAPI `entity_autocomplete` type that works like the Link field (which uses and enhances [`EntityAutocomplete`](https://git.drupalcode.org/project/drupal/-/blob/9.1.10/core/lib/Drupal/Core/Entity/Element/EntityAutocomplete.php)). Unfortunately for az_card at least, the custom field implementation means that Drupal core's nice implementation is just out of reach -- because fields cannot themselves be, well, fielded AFAIK.

Maybe that was obvious to others, but I needed to step through it. Does that all sound right so far?

Also, I have a question related to
> Some of our Forms and compound Field Widgets have a need to store links

Do you have other examples of this? I assume "compound Field Widgets" refers to `az_card`, but are there other forms that would benefit from this?

**Comment by joeparsons** [comment]
> Do you have other examples of this? I assume "compound Field Widgets" refers to `az_card`, but are there other forms that would benefit from this?

We also have an [`az_accordion`](https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_accordion/src/Plugin/Field/FieldType/AZAccordionItem.php) field.

**Comment by michaelhagedon** [comment]
I've considered a few approaches to this:
1. Extend the link module's `LinkWidget` plugin in each case of a compound field widget that needs links. This would get us access to all the protected and parent methods that core is using. The real problem with this, at least in the case of `az_card`, is that cards are not inherently links and in fact the Card Link URL field is optional. It's weird to extend a class in that case.
2. Create a new FAPI element (extended from `entity_autocomplete`) that handles external and internal urls. We discussed this a little a few weeks ago, but as I mentioned above this solution involves duplicating some of what the link module is already doing. Doing this well might be contributable. 
3. Somehow copy core's interesting link module code into a trait or something that can be shared across widgets. I didn't go far down this path, sounds complicated and still not DRY.
4. Late in my analysis, I [discovered](https://drupal.stackexchange.com/questions/213472/add-link-field-in-custom-form#answer-288280) that the [linkit module](https://www.drupal.org/project/linkit) provides this functionality. Although the stated purpose of this module is to provide better link inside _wysiwyg_ editors, the Drupal Answers post above points out that it includes a `linkit` FAPI element which handles internal and external URLs!

I'm very intrigued by option 4. I tested it out on cards simply by changing the type of the uri form element inside `AZCardWidget` and adding autocomplete as the post says:
```php
    $element['link_uri'] = [
      '#type' => 'linkit',
      '#autocomplete_route_name' => 'linkit.autocomplete',
      '#autocomplete_route_parameters' => [
        'linkit_profile_id' => 'default',
      ],
      '#title' => $this->t('Card Link URL'),
      '#element_validate' => [[$this, 'validateCardLink']],
      '#default_value' => isset($items[$delta]->link_uri) ? $items[$delta]->link_uri : NULL,
    ];
```

Here is the result of searching for existing nodes:
![image](https://user-images.githubusercontent.com/645924/125190057-1a9f9a00-e1f0-11eb-9ffa-c7dbe4a18ca3.png)

If there's no match, it assumes an external link:
![image](https://user-images.githubusercontent.com/645924/125190125-5fc3cc00-e1f0-11eb-9fd3-610df9b405fb.png)

I didn't change the AZ Card formatter or the field type, but the output is what I'd expect:
![image](https://user-images.githubusercontent.com/645924/125190209-d3fe6f80-e1f0-11eb-9c45-f272529b94f2.png)

(Those links go where they're supposed to.)

So I think the linkit approach is very promising. Advantages:
1. It does what we want!
2. It doesn't need much code of our own (so far, deeper investigation is needed to make sure this has no adverse affects).
3. The module appears to be widely used (115K installs) and I've heard of it...

Disadvantages:
1. The stated purpose of this module is not what we're using it for, and what we're using it for isn't listed as a feature.
2. It looks somewhat different from core's `entity_autocomplete` type (which linkit does not use).
3. It adds another contrib module (not necessarily negative, just more dependencies).
4. Though the code is minimal, there is a little boilerplate (the autocomplete properties). If that would always be consistent, we might be able to extend the linkit element to make those the default.

On balance, I think linkit is a good approach, but I'm open to feedback. If I don't hear anything soon, I'll work on a PR after some more testing.

**Comment by tadean** [comment]
It does seem like this would be similiar to what we want. A few observations:

- If we are using the module in a semi-unintended way, it does seem appealing to see if we can copy the FAPI widget into our codebase somehow.
- No stable release as of yet, and it's been a while since the latest beta release - are there are any pending issues we need to look out for that might be blockers?

**Comment by camikazegreen** [comment]
It looks like we may also like to use the Linkit module in the intended way, however we have some concerns over the potential for it to slow down the editor experience. Can you tell how many times that js would be loaded if we had dozens of Linkit fields on the page at once?

**Comment by bberndt-uaz** [comment]
A client wanted to link a card button to a PDF on their site but was unable to enter a relative path to the file (e.g. `/sites/default/files/<yyyy-mm>/<filename.pdf>`) in the Card Link URL field. Since the site is still under development, using an absolute URL with the current pantheonsite.io domain would cause the link to break after the site is launched.

@danahertzberg The merged PR #636 from issue #537 allowed for the use of internal paths to nodes (such as `/<node-title>` and `/node/2`) as well as the special paths `<front>` and `<none>`. However, internal paths to files such as PDFs and images do not work--the edit page displays an error and prevents the form from being saved.

**Comment by trackleft** [comment]
FYI https://www.drupal.org/project/linkit/issues/2925828
